File: sdks/rust/src/http.rs
//! The transport: headers, retries, timeouts.

use std::sync::{Arc, RwLock};
use std::time::Duration;

use serde::de::DeserializeOwned;

use crate::error::{api_error, Error, Result};

pub(crate) const KEY_HEADER: &str = "X-Atlas-Key";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Method {
    Get,
    Post,
    Delete,
}

impl Method {
    /// Whether a request may be replayed after a failure.
    ///
    /// GET and DELETE are idempotent by definition. POST is not, and this
    /// SDK will not retry one — with a single exception the caller opts
    /// into by supplying an idempotency key, which is what makes the
    /// replay safe on the server side.
    fn idempotent(self) -> bool {
        matches!(self, Method::Get | Method::Delete)
    }

    fn as_reqwest(self) -> reqwest::Method {
        match self {
            Method::Get => reqwest::Method::GET,
            Method::Post => reqwest::Method::POST,
            Method::Delete => reqwest::Method::DELETE,
        }
    }
}

pub(crate) struct Request {
    pub method: Method,
    pub path: String,
    pub query: Vec<(String, String)>,
    pub body: Option<serde_json::Value>,
    /// Set when the body carries an idempotency key, which makes a POST
    /// safe to replay.
    pub idempotent_post: bool,
    pub authenticated: bool,
}

impl Request {
    pub fn get(path: impl Into<String>) -> Self {
        Self::new(Method::Get, path)
    }

    pub fn post(path: impl Into<String>) -> Self {
        Self::new(Method::Post, path)
    }

    pub fn delete(path: impl Into<String>) -> Self {
        Self::new(Method::Delete, path)
    }

    fn new(method: Method, path: impl Into<String>) -> Self {
        Self {
            method,
            path: path.into(),
            query: Vec::new(),
            body: None,
            idempotent_post: false,
            // Most routes need one. The three that do not opt out
            // explicitly, so forgetting is a 401 rather than an
            // accidentally anonymous request.
            authenticated: true,
        }
    }

    pub fn anonymous(mut self) -> Self {
        self.authenticated = false;
        self
    }

    pub fn json(mut self, body: serde_json::Value) -> Self {
        self.body = Some(body);
        self
    }

    pub fn replayable(mut self) -> Self {
        self.idempotent_post = true;
        self
    }

    pub fn query(mut self, key: &str, value: impl ToString) -> Self {
        self.query.push((key.to_string(), value.to_string()));
        self
    }

    pub fn maybe_query(self, key: &str, value: Option<impl ToString>) -> Self {
        match value {
            Some(v) => self.query(key, v),
            None => self,
        }
    }
}

pub(crate) struct Http {
    client: reqwest::Client,
    base_url: String,
    project_key: String,
    max_retries: u32,
    token: Arc<RwLock<Option<String>>>,
}

impl Http {
    pub fn new(
        base_url: String,
        project_key: String,
        timeout: Duration,
        max_retries: u32,
        token: Arc<RwLock<Option<String>>>,
    ) -> Result<Self> {
        let client = reqwest::Client::builder()
            .timeout(timeout)
            .build()
            .map_err(|e| Error::InvalidRequest(format!("could not build HTTP client: {e}")))?;
        Ok(Self {
            client,
            base_url: base_url.trim_end_matches('/').to_string(),
            project_key,
            max_retries,
            token,
        })
    }

    pub async fn send<T: DeserializeOwned>(&self, req: Request) -> Result<T> {
        let body = self.send_raw(req).await?;
        // An empty 2xx body is legal for endpoints that return nothing.
        // `serde_json::from_str::<()>("")` fails, so null stands in.
        let text = if body.trim().is_empty() {
            "null"
        } else {
            &body
        };
        serde_json::from_str(text).map_err(|e| Error::Decode(e.to_string()))
    }

    async fn send_raw(&self, req: Request) -> Result<String> {
        let url = format!("{}{}", self.base_url, req.path);
        let attempts = if req.method.idempotent() || req.idempotent_post {
            self.max_retries + 1
        } else {
            1
        };

        let mut last: Option<Error> = None;
        for attempt in 0..attempts {
            if attempt > 0 {
                // Exponential backoff with jitter. Without jitter, every
                // client that failed together retries together, and the
                // recovering service is hit by a synchronised wave.
                let base = 100u64 << (attempt - 1);
                let jitter = fastrand_u64() % (base / 2 + 1);
                tokio::time::sleep(Duration::from_millis(base + jitter)).await;
            }

            let mut builder = self
                .client
                .request(req.method.as_reqwest(), &url)
                // Sent on every request, including register and login:
                // creating a user means creating them in a project.
                .header(KEY_HEADER, &self.project_key)
                .query(&req.query);

            if req.authenticated {
                if let Some(token) = self.token.read().expect("token lock poisoned").clone() {
                    builder = builder.bearer_auth(token);
                }
            }
            if let Some(ref body) = req.body {
                builder = builder.json(body);
            }

            match builder.send().await {
                Ok(response) => {
                    let status = response.status();
                    let text = response.text().await.unwrap_or_default();
                    if status.is_success() {
                        return Ok(text);
                    }
                    let err = api_error(status.as_u16(), &text);
                    if !err.is_retryable() || attempt + 1 == attempts {
                        return Err(err);
                    }
                    last = Some(err);
                }
                Err(e) => {
                    let err = if e.is_timeout() {
                        Error::Connection(format!("timed out after {:?}", e.url()))
                    } else {
                        Error::Connection(e.to_string())
                    };
                    if attempt + 1 == attempts {
                        return Err(err);
                    }
                    last = Some(err);
                }
            }
        }
        Err(last.unwrap_or_else(|| Error::Connection("no attempt was made".into())))
    }
}

/// A tiny xorshift, to avoid a dependency for jitter.
///
/// Not cryptographic and does not need to be: it decorrelates retry
/// timing, and nothing about that is a secret.
fn fastrand_u64() -> u64 {
    use std::cell::Cell;
    use std::time::{SystemTime, UNIX_EPOCH};
    thread_local! {
        static STATE: Cell<u64> = const { Cell::new(0) };
    }
    STATE.with(|s| {
        let mut x = s.get();
        if x == 0 {
            x = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .map(|d| d.as_nanos() as u64)
                .unwrap_or(0x2545_F491_4F6C_DD1D)
                | 1;
        }
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        s.set(x);
        x
    })
}

File: sdks/dart/lib/src/http.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'errors.dart';

const String keyHeader = 'X-Atlas-Key';

/// The transport: headers, retries, timeouts.
///
/// Built on `dart:io`'s HttpClient rather than package:http so the SDK has
/// no runtime dependencies — a package that pulls in `http` forces that
/// version choice on every Flutter app that adopts it.
class Http {
  Http({
    required String baseUrl,
    required this.projectKey,
    required this.timeout,
    required this.maxRetries,
    HttpClient? client,
  })  : baseUrl = baseUrl.endsWith('/')
            ? baseUrl.substring(0, baseUrl.length - 1)
            : baseUrl,
        _client = client ?? HttpClient();

  final String baseUrl;
  final String projectKey;
  final Duration timeout;
  final int maxRetries;
  final HttpClient _client;
  final Random _random = Random();

  String? token;

  void close() => _client.close(force: true);

  /// Send a request, retrying only when it is safe to.
  ///
  /// [replayable] marks a POST the caller has made idempotent by
  /// supplying an idempotency key. Without it a POST is sent exactly once:
  /// replaying a non-idempotent write is how one deposit becomes two.
  Future<dynamic> send({
    required String method,
    required String path,
    Map<String, dynamic>? body,
    Map<String, String>? query,
    bool authenticated = true,
    bool replayable = false,
  }) async {
    final idempotent = method == 'GET' || method == 'DELETE' || replayable;
    final attempts = idempotent ? maxRetries + 1 : 1;

    Object? lastError;
    for (var attempt = 0; attempt < attempts; attempt++) {
      if (attempt > 0) {
        // Exponential backoff with jitter. Without jitter every client
        // that failed together retries together, and the recovering
        // service is hit by a synchronised wave.
        final base = 100 * (1 << (attempt - 1));
        await Future<void>.delayed(
          Duration(milliseconds: base + _random.nextInt(base ~/ 2 + 1)),
        );
      }

      try {
        return await _attempt(
          method: method,
          path: path,
          body: body,
          query: query,
          authenticated: authenticated,
        );
      } on AtlasError catch (e) {
        if (!e.isRetryable || attempt + 1 == attempts) rethrow;
        lastError = e;
      } on AtlasConnectionError catch (e) {
        if (attempt + 1 == attempts) rethrow;
        lastError = e;
      }
    }
    throw lastError ?? AtlasConnectionError('no attempt was made');
  }

  Future<dynamic> _attempt({
    required String method,
    required String path,
    Map<String, dynamic>? body,
    Map<String, String>? query,
    required bool authenticated,
  }) async {
    final uri = Uri.parse('$baseUrl$path').replace(
      queryParameters: (query == null || query.isEmpty) ? null : query,
    );

    HttpClientResponse response;
    String text;
    try {
      final request = await _client.openUrl(method, uri).timeout(timeout);

      // Sent on every request, including register and login: creating a
      // user means creating them in a project.
      request.headers.set(keyHeader, projectKey);
      if (authenticated && token != null) {
        request.headers.set(HttpHeaders.authorizationHeader, 'Bearer $token');
      }
      if (body != null) {
        request.headers.contentType = ContentType.json;
        request.write(jsonEncode(body));
      }

      response = await request.close().timeout(timeout);
      text = await response.transform(utf8.decoder).join().timeout(timeout);
    } on TimeoutException {
      throw AtlasConnectionError('request to $uri timed out after $timeout');
    } on SocketException catch (e) {
      throw AtlasConnectionError('could not reach $uri: ${e.message}');
    } on HttpException catch (e) {
      throw AtlasConnectionError('could not reach $uri: ${e.message}');
    }

    if (response.statusCode >= 200 && response.statusCode < 300) {
      if (text.trim().isEmpty) return null;
      return jsonDecode(text);
    }
    throw _toError(response.statusCode, text);
  }

  AtlasError _toError(int status, String body) {
    try {
      final decoded = jsonDecode(body);
      if (decoded is Map<String, dynamic>) {
        final envelope = decoded['error'];
        if (envelope is Map<String, dynamic>) {
          return AtlasError(
            code: AtlasErrorCode.fromWire(envelope['code'] as String?),
            message: envelope['message'] as String? ?? 'unknown error',
            status: status,
          );
        }
      }
    } on FormatException {
      // Falls through: a non-JSON body means something between the caller
      // and the gateway answered — a proxy, an ingress 502 — and the
      // status is the only reliable signal.
    }
    return AtlasError(
      code: _codeForStatus(status),
      message: body.isEmpty
          ? 'HTTP $status'
          : body.substring(0, body.length < 200 ? body.length : 200),
      status: status,
    );
  }

  AtlasErrorCode _codeForStatus(int status) => switch (status) {
        401 => AtlasErrorCode.unauthenticated,
        403 => AtlasErrorCode.permissionDenied,
        404 => AtlasErrorCode.notFound,
        409 => AtlasErrorCode.alreadyExists,
        429 => AtlasErrorCode.rateLimited,
        503 => AtlasErrorCode.unavailable,
        _ => AtlasErrorCode.unknown,
      };
}
